src/primary_backup/node.rs

```diff
@@ -121,37 +138,73 @@ impl Node {
                     };
                 }
                 Some((message, reply_sender)) = network_receiver.recv() => {
-                    info!("Received network message {}", message);
-                    reply_sender.send("ACK".to_string()).unwrap();
-                    self.handle_msg(message.clone()).await.unwrap();
-                }
-                else => {
-                    error!("node channels are closed");
+                    info!("[{}] Received network message {}",self.address, message);
+
+                    if let msg @ PrimaryAddress = &message {
```

Comment:
Good solution although this can potentially be confusing for someone who is looking at the code (because it's an edge case for network messages), so I suggest adding a comment with a brief explanation of why the solution was designed this way.

Reply:
Instead of having this as a special case, can't the reply be made a new message type? That's what I did for the case of the blockchain, a node can ask GetState and the other replies with State:

https://github.com/Entropy1729/consensus-workbench/blob/91500c174ec53292ee1d312ff0bea9b4c1c2dd94/src/blockchain/node.rs#L174-L189

Reply:
also I wonder if the "give me your primary"  could be avoided by some use of the subscribe message or by having nodes announce what their primary is